lib/pharos/salt.rb

```diff
@@ -16,54 +15,113 @@ def initialize
+      JSON.parse(res.body)["return"].first[mid]
+    end
+
+    class << self
+      # Call the orchestration
+      def orchestrate
+        perform_run_request(data: { client: "runner_async",
+                                    fun: "state.orchestrate",
+                                    kwargs: { mods: "orch.kubernetes" } })
+      end
+    end
+
+    def has_roles?(mid:)
+      !minion(mid: mid)["roles"].blank?
+    end
+
+    def map_role(role:)
+      case role
+      when :master then %i(kube-master etcd)
+      when :minion then %i(kube-minion)
+      end
+    end
+
+    def assign_role(mid:, role:)
+      append_grain mid: mid, key: "roles", val: map_role(role: role)
+    end
+
     protected
 
+    def append_grain(mid:, key:, val:)
+      perform_request(endpoint: "/minions",
+                      method: "post",
+                      data: { client: "local",
+                              tgt: mid,
+                              fun: "grains.append",
+                              kwarg: { key: key, val: val } })
+    end
+
     # Performs a log in request and sets the @token instance variable on success.
     def login!
-      hsh = {
-        username: ENV["PHAROS_SALT_USER"],
-        password: "password",
-        eauth:    "auto"
-      }
-
-      res = perform_request("/login", "post", hsh)
-      @token = res["X-Auth-Token"]
+      @token ||= perform_request(endpoint: "/login",
+                                 method: "post",
+                                 data: Salt.auth_details)["X-Auth-Token"]
     end
 
     # Performs an HTTP request to the Salt API server. This methods already sets
     # some headers like "X-Auth-Token" automatically.
     #
     # Returns the response object.
-    def perform_request(endpoint, method, data = {})
-      uri = URI.join("http://#{@hostname}", endpoint)
+    def perform_request(endpoint:, method:, data: {})
+      login! unless endpoint == "/login"
+
+      uri = URI.join("http://#{Salt.hostname}", endpoint)
 
       req = Net::HTTP.const_get(method.capitalize).new(uri)
       req["Accept"]       = "application/json; charset=utf-8"
-      req["Content-Type"] = "application/x-www-form-urlencoded"
+      req["Content-Type"] = "application/json; charset=utf-8"
       req["X-Auth-Token"] = @token if @token
-      req.set_form_data(data)
+      req.body = data.to_json unless data.blank?
 
       opts = { use_ssl: false, open_timeout: 2 }
       Net::HTTP.start(uri.hostname, uri.port, opts) { |http| http.request(req) }
     end
+
+    class << self
+      def auth_details
+        {
+          username: ENV["PHAROS_SALT_USER"],
+          password: ENV["PHAROS_SALT_PASSWORD"],
+          eauth:    "pam"
+        }
+      end
+
+      def hostname
+        "#{ENV["PHAROS_SALT_HOST"]}:#{ENV["PHAROS_SALT_PORT"]}"
+      end
+
+      # Perform an HTTP request to the Salt API server in a one-shot fashion.
+      def perform_run_request(data: {})
```

Comment:
This one only POSTS and won't take or generate any token, it's a bit different but we might want to merge them, yes. The one-shot one will probably only be used from static methods (no @token required or generated).